### PR TITLE
Add macro metric support

### PR DESCRIPTION
**Pull request by overvenus**
In this PR I add a feature which can log metrics by macro.

## `metric-rust`

It is in tikv now.
~~I create a new crate `metric-rust` instead of a new mod of tikv, because some macro use `$crate` and it seems that `$crate` does not work correctly if these macro was placed in a sub mod.
I have tried to create a new sub mod in `util` but compiler reports error: unresolved name `xxx` [E0425], and this error will not occur if I place these macro in a crate. Anyway, it works.~~

Basically, `metric-rust` use the way of crate `log` uses to achive static metric client. Just like `log`, we should initialize metric early in the runtime of the program. `metric-rust` requires that clients have to satisfy trait [`Metric`](https://github.com/overvenus/metric-rust/blob/master/src/lib.rs#L28)
I am **not sure** about the safetness of `metric-rust` right now, @ngaut @BusyJay if you prefer macro, would you please review my code ~~at  [metric-rust](https://github.com/overvenus/metric-rust)~~, thanks!

ps: crate `log`: https://github.com/rust-lang-nursery/log

## statsd

I have added two types of statsd client, log and udp. They wraps `cadence::StatsdClient<T>` and *redundantly*, at least looks so, implement `Metric` for some reason:
 1. we cant directly `impl XXX for cadence::StatsdClient<T> {}` outside of crate cadence, so wraps
 2. `StatsdClient<LoggingMetricSink>` and `StatsdClient<UdpMetricSink>` are two different types, actually, plus statsd API are defined separately in `Count`, `Time` , etc., so generic is too complex.

I know it is ugly, please let me know if you have any question or suggestions.

### CODE REVIEW

**Comment by BusyJay** [comment]
Please don't create a new crate. `$crate` should work perfectly, see also [The variable $crate](https://doc.rust-lang.org/book/macros.html#the-variable-crate).

**Comment by BusyJay** [comment]
Why define a `Metric` trait instead of just using all the trait `cadence` has already defined?

**Comment by overvenus** [comment]
@BusyJay

Oh, I forget append mod name after $crate, it works perfectly indeed, I will update this PR later.

---------------------

```rust
// static client in metric-rust.
static mut CLIENT: *const Metric = &NopMetric;

// defined in metric-rust
pub trait Metric: Sync + Send { }

// defined in log
pub trait Log: Sync+Send { }

// defined in cadence
pub trait MetricSink { }
```

 - CLIENT is a raw pointer to trait.
 - `MetricSink` does not inherit `Sync` and `Send`.
 - statsd APIs in cadence are defined in different traits.

**Comment by BusyJay** [comment]
You can use `Box<MetricSink + Sync + Send>`. See also [Trait Objects](https://doc.rust-lang.org/book/trait-objects.html).

**Comment by overvenus** [comment]
Sorry for the late reply

Did you mean: 
```rust
static mut CLIENT: *const cadence::MetricSink = &NopMetricSink;

pub fn set_metric<M>(make_metric: M) -> Result<(), SetMetricError>
    where M: FnOnce() -> Box<MetricSink + Sync + Send>
```

**Comment by BusyJay** [comment]
Yes, something like that.

And I don't think we need lazy evaluation here, so M can be just replaced with MetricSink box type. You can also reference util::panic_hook for how to initial and use static variables.

**Comment by overvenus** [comment]
After several try, I dont think we should choose `MetricSink` to be the static one, it will end up with creating statsd clients on the fly. 
These client both take the ownership of `MetricSink`. When I try to deref the static `MetricSink` and pass it `from_sink()`, compiler reports an error:

 > error: the trait bound `cadence::MetricSink: std::marker::Sized` is not satisfied [E0277]

```rust
#[macro_export]
macro_rules! metric_count {
    ($key:expr, $count:expr) => {{
        use cadence::{StatsdClient, Counted, NopMetricSink, MetricSink};

        let sink = $crate::util::metric::__metric_sink().unwrap_or(&NopMetricSink); // get the static client.
        let client = StatsdClient::from_sink("", *sink); // error [E0277]
        client.count($key, $count);
    }};
}
```

**Comment by BusyJay** [comment]
Oh, sorry, my mistake. What I mean should be the traits that `StatsdClient` implemented.

Check this out:

```
use cadence::{Counted, Gauged, Metered, Timed};

trait Metric: Counted + Gauged + Metered + Timed {}

impl<T: Counted + Gauged + Metered + Timed> Metric for T {}

static mut CLIENT: Option<*const Metric> = None;

macro_rules! metric_count {
    ($key:expr, $count:expr) => {
        if let Some(client) = $crate::util::metric::__client() {
            if let Err(e) = client.incr($key) {
                warn!(...)
            }
        }
    };
}
```

In this way you don't have to define `NopMetric` or implement `Metric` to several `Client<T>` redundantly.

**Comment by overvenus** [comment]
PTAL @BusyJay

File: src/bin/tikv-server.rs [hunk 1]
```diff
@@ -59,13 +62,13 @@ fn get_string_value<F>(short: &str,
...
 {
     matches.opt_str(short)
-        .or_else(|| {
-            config.lookup(long).and_then(|v| f(v)).or_else(|| {
-                info!("malformed or missing {}, use default", long);
-                default
-            })
-        })
-        .expect(&format!("please specify {}", long))
+           .or_else(|| {
```
Comment (BusyJay) [inline src/bin/tikv-server.rs]: What's your rustfmt version?

Currently we are using 0.5.0.

File: src/util/metric/mod.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+use std::fmt;
+use std::sync::atomic::{AtomicUsize, ATOMIC_USIZE_INIT, Ordering};
+
+use cadence::prelude::*;
+
+#[macro_use]
+pub mod macros;
+
+static mut CLIENT: Option<*const Metric> = None;
+static STATE: AtomicUsize = ATOMIC_USIZE_INIT;
```
Comment (BusyJay) [inline src/util/metric/mod.rs]: Why not just using an AtomicBool to indicate if `CLIENT` is initialized?

Comment (siddontang) [inline src/util/metric/mod.rs]: Can use ONCE_INIT here?

File: src/util/metric/mod.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+            return Err(SetMetricError(()));
+        }
+
+        CLIENT = Some(Box::into_raw(client));
+        Ok(())
+    }
+}
+
+#[doc(hidden)]
+pub fn __client() -> Option<&'static Metric> {
```
Comment (BusyJay) [inline src/util/metric/mod.rs]: remove the `__` prefix.

File: src/util/metric/mod.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+const INITIALIZED: usize = 1;
+
+pub trait Metric: Counted + Gauged + Metered + Timed {}
+
+impl<T: Counted + Gauged + Metered + Timed> Metric for T {}
+
+/// The type returned by `set_metric_client` if `set_metric_client` has already been called.
+#[allow(missing_copy_implementations)]
+#[derive(Debug)]
+pub struct SetMetricError(());
```
Comment (BusyJay) [inline src/util/metric/mod.rs]: Remove redundant `(())`.

File: src/util/metric/mod.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+static STATE: AtomicUsize = ATOMIC_USIZE_INIT;
+const UNINITIALIZED: usize = 0;
+const INITIALIZED: usize = 1;
+
+pub trait Metric: Counted + Gauged + Metered + Timed {}
+
+impl<T: Counted + Gauged + Metered + Timed> Metric for T {}
+
+/// The type returned by `set_metric_client` if `set_metric_client` has already been called.
+#[allow(missing_copy_implementations)]
```
Comment (BusyJay) [inline src/util/metric/mod.rs]: I think it's ok to derive `Copy`.

File: src/util/metric/mod.rs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    }
+}
+
+#[doc(hidden)]
+pub fn __client() -> Option<&'static Metric> {
+    if STATE.load(Ordering::SeqCst) != INITIALIZED {
+        return None;
+    }
+
+    unsafe { CLIENT.map_or(None, |c| Some(&*c)) }
```
Comment (BusyJay) [inline src/util/metric/mod.rs]: Just `map` is ok.

File: src/util/metric/mod.rs (hunk 2 above)
Comment (BusyJay) [inline src/util/metric/mod.rs]: No, the client is produced outside and only allow setting once.

File: src/bin/tikv-server.rs (hunk 1 above)
Comment (overvenus) [inline src/bin/tikv-server.rs]: 0.4.1, I will update it now.

**Comment by BusyJay** [comment]
LGTM

**Comment by ngaut** [comment]
PTAL @siddontang

**Comment by siddontang** [comment]
LGTM

**Comment by ngaut** [comment]
Well done!  @overvenus

**Comment by overvenus** [comment]
Thank you! @BusyJay @siddontang @ngaut

**Pull request merged**